Fixes #11371 - Review ArrayByteBufferPool eviction.

sbordet opened the pull request:
* Eviction is now performed on release(), rather than acquire().
* Memory accounting is done on release(), rather than acquire().
This is because we were always exceeding the maxMemory on acquire(), by returning a non-pooled buffer.
We only need to account for what is idle in the pool, and that is done more efficiently on release(), and it is leak-resistant (i.e. if the buffer is not returned, the memory is already non accounted for, keeping the pool consistent).
* Released entries now give precedence to Concurrent.Entry, rather than Queued.Entry, so the queued pool is always kept at minimum size.
* Eviction is performed by running through the buckets, starting at a random bucket, and removing, if present, an idle buffer until the excess memory is freed.

sbordet committed:
Fixes #11371 - Review ArrayByteBufferPool eviction.

* Eviction is now performed on release(), rather than acquire().
* Memory accounting is done on release(), rather than acquire().
This is because we were always exceeding the maxMemory on acquire(), by returning a non-pooled buffer.
We only need to account for what is idle in the pool, and that is done more efficiently on release(), and it is leak-resistant (i.e. if the buffer is not returned, the memory is already non accounted for, keeping the pool consistent).
* Released entries now give precedence to Concurrent.Entry, rather than Queued.Entry, so the queued pool is always kept at minimum size.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Restored QueuedPool.terminate() behavior.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Avoid running through the buckets to compute used memory.
Moved entry reservation to the first release().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Added eviction: one pass through all buckets to try to free excess memory.
If successful, the buffer being released is pooled, otherwise it is also discarded.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Added statistics.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Using a random index to loop over the buckets during evict().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed statistics about evict/remove in case the entry cannot be removed.
Fixed clearing of used memory in clear().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -191,40 +191,115 @@ public int getMaxCapacity()
+        return memory - maxMemory;
+    }
+
+    private boolean evict(long excessMemory, boolean direct)
+    {
+        RetainedBucket[] buckets = direct ? _direct : _indirect;
+        int length = buckets.length;
+        int index = ThreadLocalRandom.current().nextInt(length);
+        for (int c = 0; c < length; ++c)
+        {
+            RetainedBucket bucket = buckets[index++];
+            if (index == length)
+                index = 0;
+            excessMemory -= bucket.evict();
+            if (excessMemory <= 0)
+                return true;
+        }
+        return false;
```
This looks too complex.  Just pick a random bucket and if it is empty or the same size as the buffer we are releasing, then return false.    Sure it will miss a lot, but that is fine as we don't need to adapt quickly as size frequency changes are unlikely

```suggestion
    private boolean evict(long excessMemory, int releasedCapacity, boolean direct)
    {
        RetainedBucket[] buckets = direct ? _direct : _indirect;
        int length = buckets.length;
        int index = ThreadLocalRandom.current().nextInt(length);
        RetainedBucket bucket = buckets[index++];
        if (bucket == null || bucket.capacity() == releasedCapacity)
            return false;
        return bucket.evict() > excessMemory;
```

sbordet replied:
@gregw I think your suggestion is just keeping idle buffers around, because now you have a random chance to hit the right bucket (1/N) to free a truly idle buffer.

Most of the time your suggestion will fail to free enough memory, returning `false` most of the time, and therefore discarding buffers that are actually in use (because they are being released back into the pool, but they won't).

Looping over all the buckets is not "too complex", it is 5 lines of code, and I think it's more fair.

If the "small" buckets are idle because the "large" buckets are in use, there is about 50% chance to start from smaller buckets and free one buffer each, and hopefully (if enough buffers are freed) keep the one being released.

If the "small" buckets are in use, there is still 50% chance to not overgrow them, and the idle "large" buffers get a 50% chance to be hit and slowly be removed, typically pooling the one being released (because it's smaller).

gregw replied:
Wow, it's like we've swapped brains and now you are arguing my points for a loop :)

The main thing I don't like about this version is the churning. It will often evict a buffer of size X only so it can add a buffer of size X.  At the very least it should never evict from the same bucket it is trying to add to.

I see the fact that my suggestion fails to evict 1/N times as a feature. Frequently, your original PR is correct and it is best to discard the just released buffer. So I'd say my suggestion is optimal (N-1)/N of the time as it will simply act like your original PR 

Changing size frequency is not the normal case, so a 1/N adaption rate should be sufficient to move to the new size(s) whilst not having lots of threads looping.

But I'm not going to die in a ditch about this. If you really want to loop, then loop. But please avoid the churn above and consider using an Atomic so only 1 thread is looping/evicting at the same time.

gregw replied:
The other aspect of this loop is that it is biased depending on the gaps between used buckets.   If we have the following buckets:
| Bucket  | Capacity | Size |  Probability | 
| ------------- | ------------- | -----| ------ |
| 0  | 2K  |  0 |  |
| 1  | 4K  |  20 | 2/16 |
| 2  | 6K  |  0 | |
| 3  | 8K  |  0 | |
| 4  | 10K  |  0 | |
| 5  | 12K  |  0 | |
| 6  | 14K  |  0 | |
| 7  | 16K  |  20 | 6/16 |
| 8  | 18K  |  20 | 1/16 |
| 9  | 20K  |  0 | |
| 10 | 22K  |  0 | |
| 11 | 24K  |  0 | |
| 12 | 26K  |  0 | |
| 13 | 28K  |  0 | |
| 14 | 30K  |  0 | |
| 15 | 32K  |  20 | 7/16 |

Then the 32K buckets are 7 times more likely to be evicted than the 18K buckets, and 3.5 times more likely than the 4K buckets.

I'd much rather just evict 3/16 of the time and keep it unbiased.

sbordet replied:
@gregw you make good points, but I don't think we have enough data to decide.
Let's discuss this in person, and/or when we have more data.

gregw replied:
OK, but I will type a few more thoughts here so I don't forget before we can hangout.

From a small sample, we "typically" have 16 buckets of which 4 to 6 sizes are used, "typically" 2 or 3 of those have a lot more usage than the others.

If we are release a buffer that will exceed the max size then either:
 + it's bucket already has some idle buffers of the same size, so we can just discard this buffer.
 + it's bucket has no idle buffers, so we need to evict from another bucket picked at random:
    * 1/16 chance that it picks the same bucket, so give up and discard.  We might be unbalanced but chances are a future release will fix it.
    * 10/16 chance that it picks an empty bucket, so give up and discard. We might be unbalanced but chances are a future release will fix it.
    * 5/16 chance that a non empty bucket is picked, which might be:
      - not enough space is evicted for the released buffer, so clear what we can and discard.  We might be unbalanced, but have freed some space and future releases can improve.
      - enough space is evicted, so the released buffer is added and we are probably more balanced
 
So more often than not, this will just discard the buffer when max capacity is exceeded, which is simple, fast and probably the best thing to do in almost all circumstances... unless we are unbalanced due to a size frequency change... in which case there is a reasonable chance (2.5/16) that we will move towards being more balanced.  The release rate is such that we will quickly rebalance even with a small chance.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -520,18 +533,85 @@ public String toString()
+                super(strategyType, maxSize, maxMultiplex);
+            }
+
+            @Override
+            protected void onAcquired(Entry<RetainableByteBuffer> entry)
+            {
+                super.onAcquired(entry);
+                updateMemory(-_capacity, entry.getPooled().isDirect());
+            }
+
+            @Override
+            protected void onReleased(Entry<RetainableByteBuffer> entry)
+            {
+                super.onReleased(entry);
+                updateMemory(_capacity, entry.getPooled().isDirect());
+            }
```
I'd prefer the memory accounting to be done directly in the ABBP methods `acquire`, `reserve` and `release` rather than as side effects of the Pool acquire/release.

lorban replied:
+1: identical behavior could be achieved by testing the return value of acquire and release, there is no need for any new API.

sbordet replied:
Done.

gregw reviewed (changes requested):
I like evicting from a random bucket, but I think you've gone too complex and it can be simplified.  Just pick one bucket randomly and:

+ it will mostly pick an empty bucket, so it will just discard the buffer being released. In a static size frequency situation, that is the best thing to do, so we mostly do the best thing and only adapt some of the time.
+ if the random bucket picked is the same capacity as the released buffer, it can just be discarded rather than evicting another. This avoid thrashing of adding then evicting a buffer.
+ if there is a bucket with lots of idle buffers that is not being used, then about 5-10% of the time it will be picked and evict some buffers. That feels like a reasonable speed to adapt.


Also, can you add the changes from https://github.com/jetty/jetty.project/commit/59e664ce2b6d49b3982f94805b35eaba6c9337ab for CompoundPool, QueuedPool and ConcurrentPool to this PR. I know the counts are now not on any critical path, but it itches me that we will getIdleCount, if called, will stream over all entries, when it can just return size.

Should we also deprecate the Quadratic impl (make it just be the normal pool) and remove the injectable bucket calculation complexity?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -520,18 +533,85 @@ public String toString()
+                super(concurrentBucket, queuedBucket);
+            }
+
+            private Pool.Entry<RetainableByteBuffer> evict()
```
This method is pretty simple and only called from one place. It could be inlined which would save having this internal class.

sbordet replied:
I like it better in its own class.

sbordet committed:
Removed overridden methods to keep track of memory usage.
Now the memory usage is computed on-the-spot when acquiring, releasing and evicting.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Overridden get*Count() methods in Pool implementations for efficiency.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Skipping the bucket target of release during eviction.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
JMH benchmark for ABBP.

Here, testing the impact of statistics for unconstrained ABBP:

With `AtomicLong`: `5.287` vs `7.017` Mops/s
```
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity            4096               -1          65536            0              0                 true  thrpt   10  5287439.484 ±  53877.187  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity            4096               -1          65536            0              0                false  thrpt   10  7016694.751 ± 226393.784  ops/s
```

With `LongAdder`: `6.132` vs `7.044` Mops/s
```
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity            4096               -1          65536            0              0                 true  thrpt   10  6132287.148 ±  65148.418  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity            4096               -1          65536            0              0                false  thrpt   10  7043554.688 ± 194914.916  ops/s
```

sbordet committed:
Added field `statisticsEnabled`.
Using LongAdder for statistics.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
JMH benchmark for ABBP.

Here, testing @sbordet evict algorithm (1 loop over all buckets starting at random index, skip same capacity bucket) vs @gregw algorithm (pick 1 random bucket, skip same capacity bucket, evict from bucket):

@sbordet's:
```
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity            4096               -1          65536            0              0                 true  thrpt   10  9002854.971 ± 101884.524  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity            4096               -1          65536      1048576              0                 true  thrpt   10  4380148.459 ± 325418.623  ops/s

org.eclipse.jetty.io.ArrayByteBufferPool@26edac07{min=0,max=65536,buckets=16,heap=0/1048576,direct=1048576/1048576}
+> direct size=16
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@30853dbf{capacity=4096,in-use=0/0,pooled/acquires=2457205/2659905(92.379%),non-pooled/evicts/removes/releases=5169/51188/197531/2659905}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@22ecfe8a{capacity=8192,in-use=0/11,pooled/acquires=45111533/45216172(99.769%),non-pooled/evicts/removes/releases=4556/2274284/100072/45216172}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@3ab1e25f{capacity=12288,in-use=0/0,pooled/acquires=2216231/2659083(83.346%),non-pooled/evicts/removes/releases=28963/49987/413889/2659083}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@23404873{capacity=16384,in-use=0/0,pooled/acquires=2229354/2661871(83.751%),non-pooled/evicts/removes/releases=38418/51400/394099/2661871}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@63b908c2{capacity=20480,in-use=0/0,pooled/acquires=2245768/2659630(84.439%),non-pooled/evicts/removes/releases=52350/52598/361512/2659630}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@2664a291{capacity=24576,in-use=0/3,pooled/acquires=2271665/2657337(85.487%),non-pooled/evicts/removes/releases=62376/53429/323293/2657337}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@13721f18{capacity=28672,in-use=0/3,pooled/acquires=2299657/2661455(86.406%),non-pooled/evicts/removes/releases=72219/54659/289576/2661455}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@32d0ded5{capacity=32768,in-use=0/1,pooled/acquires=2313063/2658056(87.021%),non-pooled/evicts/removes/releases=81802/54701/263190/2658056}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@19043f55{capacity=36864,in-use=0/1,pooled/acquires=2326751/2661217(87.432%),non-pooled/evicts/removes/releases=90917/55126/243548/2661217}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@1a7766ce{capacity=40960,in-use=0/1,pooled/acquires=2325787/2656036(87.566%),non-pooled/evicts/removes/releases=100901/55184/229347/2656036}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@19fe3047{capacity=45056,in-use=0/2,pooled/acquires=2331023/2661338(87.588%),non-pooled/evicts/removes/releases=111051/56061/219262/2661338}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@2d6c63db{capacity=49152,in-use=0/4,pooled/acquires=2325386/2659990(87.421%),non-pooled/evicts/removes/releases=123582/56956/211018/2659990}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@15aa1fb4{capacity=53248,in-use=0/2,pooled/acquires=2315380/2660379(87.032%),non-pooled/evicts/removes/releases=138286/56069/206711/2660379}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@344a5ad9{capacity=57344,in-use=0/3,pooled/acquires=2304223/2662086(86.557%),non-pooled/evicts/removes/releases=155233/56461/202627/2662086}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@5cde281f{capacity=61440,in-use=0/2,pooled/acquires=2283099/2658423(85.882%),non-pooled/evicts/removes/releases=175578/55870/199744/2658423}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@1262fae4{capacity=65536,in-use=0/0,pooled/acquires=2264663/2661456(85.091%),non-pooled/evicts/removes/releases=201161/55321/195632/2661456}
```

@gregw's:
```
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity            4096               -1          65536            0              0                 true  thrpt   10  9097094.184 ± 146472.471  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity            4096               -1          65536      1048576              0                 true  thrpt   10  4559670.608 ± 376450.879  ops/s

org.eclipse.jetty.io.ArrayByteBufferPool@6f3b9201{min=0,max=65536,buckets=16,heap=0/1048576,direct=1011712/1048576}
+> direct size=16
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@28583117{capacity=4096,in-use=0/2,pooled/acquires=2575760/2786198(92.447%),non-pooled/evicts/removes/releases=6450/57604/203986/2786198}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@2362f293{capacity=8192,in-use=0/10,pooled/acquires=46391093/47336444(98.003%),non-pooled/evicts/removes/releases=57604/2460704/887737/47336444}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@69551d3e{capacity=12288,in-use=0/1,pooled/acquires=2562090/2784290(92.020%),non-pooled/evicts/removes/releases=15129/60100/207070/2784290}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@ec598a{capacity=16384,in-use=0/2,pooled/acquires=2556711/2785834(91.775%),non-pooled/evicts/removes/releases=21557/60207/207564/2785834}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@48e32739{capacity=20480,in-use=0/2,pooled/acquires=2543488/2783059(91.392%),non-pooled/evicts/removes/releases=31322/60665/208247/2783059}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@7c21a2dc{capacity=24576,in-use=0/0,pooled/acquires=2533658/2782756(91.049%),non-pooled/evicts/removes/releases=41266/61963/207832/2782756}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@3ddcec8b{capacity=28672,in-use=0/3,pooled/acquires=2521220/2784393(90.548%),non-pooled/evicts/removes/releases=54347/61797/208823/2784393}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@5f1ed2b3{capacity=32768,in-use=0/2,pooled/acquires=2506601/2782996(90.068%),non-pooled/evicts/removes/releases=67510/61721/208883/2782996}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@e15ec86{capacity=36864,in-use=0/1,pooled/acquires=2494442/2785667(89.546%),non-pooled/evicts/removes/releases=82498/62708/208726/2785667}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@7fd388da{capacity=40960,in-use=0/2,pooled/acquires=2480469/2785648(89.045%),non-pooled/evicts/removes/releases=97254/61994/207923/2785648}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@25318b1f{capacity=45056,in-use=0/1,pooled/acquires=2461254/2784967(88.376%),non-pooled/evicts/removes/releases=114958/61722/208754/2784967}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@c134e40{capacity=49152,in-use=0/1,pooled/acquires=2443864/2784654(87.762%),non-pooled/evicts/removes/releases=132877/62006/207912/2784654}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@2d43f298{capacity=53248,in-use=0/2,pooled/acquires=2423021/2781587(87.109%),non-pooled/evicts/removes/releases=151559/61289/207005/2781587}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@4e352ab2{capacity=57344,in-use=0/3,pooled/acquires=2407192/2788218(86.334%),non-pooled/evicts/removes/releases=174637/60987/206386/2788218}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@67f53918{capacity=61440,in-use=0/1,pooled/acquires=2378839/2784073(85.445%),non-pooled/evicts/removes/releases=199095/61405/206138/2784073}
|  +> org.eclipse.jetty.io.ArrayByteBufferPool$RetainedBucket@71820ae8{capacity=65536,in-use=0/2,pooled/acquires=2352840/2784936(84.485%),non-pooled/evicts/removes/releases=227259/61391/204835/2784936}
```

@sbordet's is slightly worse in throughput, but better at hitRatio on the buffers actually in use.

For larger input buffers (the 8 KiB above), @sbordet's keeps a high hitRatio, but @gregw's become way worse.
For example, with 60 KiB input buffer, @sbordet's hitRatio on that buffer is 99.94% at 3045807 ops/s, while @gregw's hitRatio is 84.9% at 1642158 ops/s.
The buffer currently released is almost always pooled in @sbordet's, but almost always discarded in @gregw's, causing re-allocation (whose main cost is zeroing the buffer).

sbordet committed:
Added percentage to statistics reporting.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Added JMH benchmark.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented:
@sbordet interesting!  So it is worthwhile spending some time looping to find a bucket that can be evicted so that the just released buffer can be pooled.   I wonder how my most idle bucket algorithm would have gone?

I've got a couple of ideas for improving the loop to make it less biased... stand by....

gregw commented:
@sbordet I think we are really close now... and probably over-analysing beyond the fidelity of JMH.... but we are here now, so let's finish...

Your JMH test is OK, but by spreading the random writes over the full range, it is not like what we have seen with many voids in the buckets, that will affect the bias of the current algorithm.    So I've made an attempt to alter the JMH so that it uses a fixed set of random write sizes, thus there should be some empty buckets.

I've also done two other eviction algorithms: 
 + `evictUnbiasedRandom`, that needs to iterate twice so it can count the non empty buckets (of course that is racy but it is an attempt).
 + `evictLargest`, my original algorithm to find the largest bucket, which of course does nested iterations, but only on non-empty buckets

For some reason, my setup is not able to run the JMHs at the moment (plus I've only got my pathetic laptop).      So can you run these to compare.

Once you have the numbers, let's hangout with @lorban and pick one.

gregw commented:
@sbordet my changes are in https://github.com/jetty/jetty.project/tree/fix/jetty-12/11371/review-abbp-eviction3  I can push to your branch if you like?

sbordet commented:
> Your JMH test is OK, but by spreading the random writes over the full range, it is not like what we have seen with many voids in the buckets, that will affect the bias of the current algorithm.

What we have seen so far were CometD benchmarks over HTTP/2, using ServletOutputStream.

The moment we use a `Writer` for HTTP, or we use WebSocket, the range of buffers used will depend entirely on the application writes, which may be clustered around certain sizes, but we do not know.

gregw commented:
> The moment we use a `Writer` for HTTP, or we use WebSocket, the range of buffers used will depend entirely on the application writes, which may be clustered around certain sizes, but we do not know.

We do know that many frameworks like cometd, servlets, etc. will impose some level of buffering/aggregation before the writes, so quantization of sizes is definitely a common use case, but perhaps not the only one.     Hence my JMH modifications make the writeDiversity a parameter that we can adjust to see the different effects of voids versus no voids.

As I said, we are probably close to beyond the fidelity of JMH to simulate reality, but modelling the main use case that we have seen is certainly something we should do.

sbordet commented:
Agreed, I'll prepare some number for this afternoon.

sbordet committed:
Disabled statistics by default.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -191,40 +202,128 @@ public int getMaxCapacity()
+        RetainableByteBuffer buffer = entry.getPooled();
+        BufferUtil.reset(buffer.getByteBuffer());
+        boolean direct = buffer.isDirect();
+        long excessMemory = getExcessMemory(bucket, direct);
+        if (excessMemory > 0)
```
This algorithm is racy and it cannot guarantee that the max capacity is respected.  Multiple threads might all calculate excess memory at the same time and all get an answer saying they are 1 byte below max.  So all  threads will not evict and all 10 will updateMemory below.

I think we should do an atomicUpdate here and return the excess from that.   We can then do the evicting and if that is not enough discard the buffer and update the memory again.

sbordet replied:
@gregw turns out there always is a risk of over-shooting the max memory.

We can precisely update the memory when it's within maxMemory; but when the update would overshoot we would not update the memory, so another thread could see a stale value, because the first thread would now try to evict and free space.

Seems too complicated to try to be precise, so the current compromise is simple and speedy but not precise.

gregw replied:
@sbordet I know we cannot fix the race entirely.  But we can use the atomic that we already have to do a bit better than either having multiple threads not evicting when they should or multiple threads evicting when they should not.

The max memory is already atomic, so why not updating it atomically so that each thread knows its contribution to the amount over the max and then proceeds to try to evict only that much memory.

Even if it doesn't solve the race, it will be faster because we will not access the atomic twice (once to test and once to update).

Standby, I'll prepare a suggestion for this....

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/QueuedPool.java:
```diff
@@ -116,6 +117,14 @@ public Entry<P> acquire()
+    protected void onReleased(Entry<P> entry)
+    {
+    }
+
```
These are no longer needed

sbordet committed:
Updates after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Removed event methods for acquire and remove.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Added statisticsEnabled property in Jetty module bytebufferpool.mod.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
